### PR TITLE
[Baremetal] Add FI/SCA compliant versions of mem-functions

### PR DESCRIPTION
**Pull request by jarlamsa**
## Description
Add FI/SCA compliant versions for memcmp, memcpy and memset. These functions should be used in global scope wherever mem-functions are used with secrets.
Adhering to the standards requires also a global RNG that can be used inside the global functions.

## Status
**REVIEW**

Impact on size with the baremetal config with ARM-GCC_6.3.1 and ARMC5_5.06 by running `scripts/baremetal.sh --rom --gcc --armc5`

With `MBEDTLS_ENTROPY_HARDWARE_ALT` enabled

-- | ARM-GCC | ARMCC-5
-- | -- | --
Before | 42191 | 43428
After | 42287 | 43576
Gained | 94 | 148

### CODE REVIEW

File: include/mbedtls/platform_util.h [hunk 1]
```diff
@@ -164,6 +164,14 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
  */
 void mbedtls_platform_zeroize( void *buf, size_t len );
 
+void mbedtls_platform_memset( void *ptr, int value, size_t num);
```
Comment (jarvte) [inline include/mbedtls/platform_util.h]: Minor, coding conventions: add space after num

File: include/mbedtls/platform_util.h [hunk 2]
```diff
@@ -164,6 +164,14 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
...
+
+void mbedtls_platform_memcpy( void *dst, const void *src, size_t num );
+
+int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num);
```
Comment (jarvte) [inline include/mbedtls/platform_util.h]: Minor, coding conventions: add space after num

File: include/mbedtls/platform_util.h [hunk 3]
```diff
@@ -164,6 +164,14 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
  */
 void mbedtls_platform_zeroize( void *buf, size_t len );
 
+void mbedtls_platform_memset( void *ptr, int value, size_t num);
+
+void mbedtls_platform_memcpy( void *dst, const void *src, size_t num );
```
Comment (jarvte) [inline include/mbedtls/platform_util.h]: Please add doxygen description for all methods

File: library/platform_util.c [hunk 4]
```diff
@@ -79,6 +79,70 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
...
+        return 0;
+    }
+
+    return 1;
+}
+
+//TODO: This is a stub implementation of the global RNG function.
+size_t mbedtls_random_in_range( size_t num )
+{
+    return 0;
```
Comment (jarvte) [inline library/platform_util.c]: add for example `( void ) num;` to get it pass the ci

File: include/mbedtls/platform_util.h (hunk 3 above)
Comment (jarlamsa) [reply include/mbedtls/platform_util.h]: Will be done in a later commit, this PR is under development.

File: library/platform_util.c [hunk 5]
```diff
@@ -79,6 +79,71 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
...
+void mbedtls_platform_memset( void *ptr, int value, size_t num )
+{
+    /* Randomize start offset. */
+    size_t startOffset = mbedtls_random_in_range( num );
+    /* Randomize data */
+    size_t data = mbedtls_random_in_range( 0xff );
+
+    /* Perform a pair of memset operations from random locations with
+     * random data */
+    memset( ( void * ) ( ptr + startOffset ), value, ( num - startOffset ) );
```
Comment (artokin) [inline library/platform_util.c]: ARMC5 compiler gives "_platform_util.c", line 91: Error:  #852: expression must be a pointer to a complete object type"_. Would you please add type cast like already done in other functions?

Comment (jarlamsa) [reply library/platform_util.c]: Done

**Comment by jarlamsa** [comment]
Thank you for the review-comments @mpg I think those have now been addressed. Tests and the 0-length corner cases needs still to be documented in doxygen.

**Comment by mpg** [comment]
@jarlamsa I just noticed that the function prototype don't exactly match those of the standard function. Would you consider cherry-picking the following commit to this PR? https://github.com/ARMmbed/mbedtls/pull/2877/commits/f81e15895d223f87f66b3e68af86d5f7dce1c203 Thanks!

**Comment by jarlamsa** [comment]
Thanks @mpg for noticing that and providing a fix.

File: library/platform_util.c [hunk 6]
```diff
@@ -135,13 +135,20 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
...
+#if !defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
     (void) num;
     return 0;
+#else
+    uin32_t result = 0;
+    size_t olen = 0;
+
+    mbedtls_hardware_poll( NULL, (unsigned char *) &result, sizeof( result ),
+                           &olen );
+    return( result );
```
Comment (mpg) [inline library/platform_util.c]: I assume you mean `result % num`?

Comment (jarlamsa) [reply library/platform_util.c]: Yes

File: include/mbedtls/platform_util.h [hunk 7]
```diff
@@ -164,6 +164,77 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
...
+/**
+ * \brief       Secure memset
+ *
+ *              This function is meant to provide a more secure way to do
+ *              memset. It starts by initialising the given memory area
+ *              from random tail location with random data. After tail is
+ *              initialised, the remaining head of the buffer is initialised
+ *              with random data. After initialisation, the original memset
+ *              is performed
+ *
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply include/mbedtls/platform_util.h]: Actually it is a constant time memcmp without the `MBEDTLS_ENTROPY_HARDWARE_ALT` only the randomness is missing without that.

File: include/mbedtls/platform_util.h [hunk 8]
```diff
@@ -164,6 +164,77 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
...
+ * \param ptr   Buffer to be set.
+ * \param value Value to be used when setting the buffer.
+ * \param num   The length of the buffer in bytes.
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply include/mbedtls/platform_util.h]: Actually I don't think we should be doing any extra branching here. 0 length for the mem-functions is specified in the standard.

**Comment by jarlamsa** [comment]
> Why copy-paste mbedtls_hardware_poll, can't you put it one common file? Optional fix.

I don't think there is such a place currently for this kind of external dependency, otherwise would be a good solution to have some utility-header for hardware-functions.

File: include/mbedtls/platform_util.h [hunk 9]
```diff
@@ -164,6 +164,83 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
...
+ * \param dst   Destination buffer where the data is being copied to.
+ * \param src   Source buffer where the data is being copied from.
+ * \param num   The length of the buffers in bytes.
+ *
+ * \return      The value of \p dst.
+ */
+void *mbedtls_platform_memcpy( void *dst, const void *src, size_t num );
+
+/**
+ * \brief       Secure memcmp
```
Comment (mpg) [inline include/mbedtls/platform_util.h]: I just noticed: this is actually not a real equivalent of `memcmp()`, as the sign of the result we return is not significant. This should probably be noted explicitly.

File: include/mbedtls/platform_util.h [hunk 10]
```diff
@@ -164,6 +164,83 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
...
+ *              through the tail part of buffers first byte by byte. After
+ *              that it starts going through the head part of buffer. In the
+ *              end, the number of equal bytes is compared to the length of the
+ *              buffers, thus making the function a fixed time memcmp.
+ *
+ * \param buf1  First buffer to compare.
+ * \param buf2  Second buffer to compare against.
+ * \param num   The length of the buffers in bytes.
+ *
+ * \return      0 if the buffers were equal.
```
Comment (mpg) [inline include/mbedtls/platform_util.h]: I suggest adding: or an unspecified non-zero value otherwise, the sign of which is not significant.

File: include/mbedtls/platform_util.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/platform_util.h]: Good point.

By the way, while you're editing the documentation, I didn't mention it earlier, but I think the description may be a bit too low-level. I'd suggest something like (for memcmp for example): "This is a constant-time version of memcmp(). If `MBEDTLS_ENTROPY_HARDWARE_ALT` is defined, the order is also randomized using the hardware RNG in order to further harden against side-channel attacks." This gives us more freedom to evolve the implementation later if needed.

File: include/mbedtls/platform_util.h (hunk 10 above)
Comment (jarlamsa) [reply include/mbedtls/platform_util.h]: Mentioning the sign might be confusing, I think the "unspecified non-zero value otherwise" is strong enough statement.

Comment (mpg) [reply include/mbedtls/platform_util.h]: Fine with me!

File: include/mbedtls/platform_util.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (mpg) [reply include/mbedtls/platform_util.h]: Good point. I withdraw that comment.

File: include/mbedtls/platform_util.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jarlamsa) [reply include/mbedtls/platform_util.h]: Good suggestion, I'll use that for these.

File: library/platform_util.c (hunk 5 above)
Comment (mpg) [reply library/platform_util.c]: I verified this is done, and the CI is happy with the fix as well, so I'm marking the conversation as resolved in @artokin's absence.

File: library/platform_util.c [hunk 11]
```diff
@@ -79,6 +83,87 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
 }
 #endif /* MBEDTLS_PLATFORM_ZEROIZE_ALT */
 
+void *mbedtls_platform_memset( void *ptr, int value, size_t num )
```
Comment (sbutcher-arm) [inline library/platform_util.c]: Architecture - to partially answer my own question above, because we're overwriting with random data, rather than name the function `mbedtls_platform_memset()`, maybe it should be named `mbedtls_platform_erase_secret()`, and we should abstract away `memset` to allow the target to provide its own `memset`.

File: library/platform_util.c [hunk 12]
```diff
@@ -79,6 +83,87 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
...
+     * random data */
+    memset( (void *) ( (unsigned char *) ptr + start_offset ), data,
+            ( num - start_offset ) );
+    memset( (void *) ptr, data, start_offset );
+
+    /* Perform the original memset */
+    return( memset( ptr, value, num ) );
+}
+
+void *mbedtls_platform_memcpy( void *dst, const void *src, size_t num )
```
Comment (sbutcher-arm) [inline library/platform_util.c]: Architecture - same point about target abstraction as for above.

File: library/platform_util.c [hunk 13]
```diff
@@ -79,6 +83,87 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
...
+    uint32_t data = mbedtls_platform_random_in_range( 256 );
+
+    memset( (void *) dst, data, num );
+    memcpy( (void *) ( (unsigned char *) dst + start_offset ),
+            (void *) ( (unsigned char *) src + start_offset ),
+            ( num - start_offset ) );
+    return( memcpy( (void *) dst, (void *) src, start_offset ) );
+}
+
+int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num )
```
Comment (sbutcher-arm) [inline library/platform_util.c]: Architecture - same point about target abstraction as for above.

File: programs/aes/aescrypt2.c [hunk 14]
```diff
@@ -80,6 +80,18 @@ int main( void )
 }
 #else
 
+#if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
```
Comment (sbutcher-arm) [inline programs/aes/aescrypt2.c]: Question - Why were all these additions to the samples necessary?

This should only be needed in an Mbed OS build, and in that we won't build the `programs`.

Comment (jarlamsa) [reply programs/aes/aescrypt2.c]: To make CI happy when `MBEDTLS_ENTROPY_HARDWARE_ALT` is defined. (The CMake build builds also the programs.)

**Comment by sbutcher-arm** [comment]
After reviewing this PR with @mpg, I've decided to promote it back to 'ready for merge'.

I do think rework (or further review and debate) is required before upstreaming but for `baremetal` this PR is acceptable.
